core/frontend/src/tile/IDBTileStorage.ts

```diff
@@ -0,0 +1,182 @@
+      if (storedResponse.result !== undefined) {
+        console.log("Stored Response found.");
+
+        // if the content has an expiration time
+        if (this._expirationTime) {
+          // We want to know when the result was stored, and how long it's been since that point
+          const timeSince = Date.now() - storedResponse.result.timeOfStorage;
+          console.log("Time Since Storage: ", timeSince / 1000, " secs" );
+
+          // If it's been greater than our time limit, delete it and return undefined.
+          if (timeSince > this._expirationTime) {
+            console.log("Stored Response Expired, Deleting content");
+            await this.deleteContent(uniqueId);
+            return undefined;
+          }
+        }
+        const content = storedResponse.result.content;
+        console.log("Returning the following content: ");
+        console.log(content);
+        await this.close();
+        return content;
+
+      } else {
+        console.log("No matching results found in db");
+      }
+      await this.close();
+      return undefined;
+    };
+    return undefined;
+  }
+
+  public async deleteContent(uniqueId: string) {
+    const deleteTransaction = await this._db.transaction("cache", "readwrite");
+    const requestDelete = await deleteTransaction.objectStore("cache").delete(uniqueId);
+
+    requestDelete.onsuccess = () => {
+      console.log("EXPIRED RESPONSE DELETED.");
+    };
+
+    deleteTransaction.onsuccess = () => {
+      console.log("DELETE TRANSACTION SUCCESS");
+    };
+
+    deleteTransaction.oncomplete = async () => {
+      console.log("DELETE TRANSACTION COMPLETED");
+    };
+  }
+
+  public async addContent(uniqueId: string, content: ArrayBuffer) {
+    await this.open();
+    const addTransaction = await this._db.transaction("cache", "readwrite");
+    const objectStore = await addTransaction.objectStore("cache");
+
+    const data = {
+      uniqueId,
+      content,
+      timeOfStorage: Date.now(),
+    };
+
+    console.log("ADDING THIS CONTENT TO THE DB");
+    console.log(data);
+
+    const requestAdd = await objectStore.add(data);
+    requestAdd.onsuccess = () => {
+      console.log("ADD REQUEST SUCCESS");
+    };
+
+    addTransaction.onsuccess = () => {
+      console.log("WRITE TRANSACTION SUCCESS");
+    };
+
+    addTransaction.oncomplete = async () => {
+      console.log("WRITE TRANSACTION COMPLETE");
+      await this.close();
+    };
+  }
+}
+
+export class IDBTileRequestChannel extends TileRequestChannel {
+  private _tileStorage: IDBStorage;
+
+  public constructor(channelName: string, cacheConcurrency: number, dbName: string) {
+
+    console.log("CREATING CHANNEL: ", channelName);
+    super(channelName, cacheConcurrency);
+
+    this._tileStorage = new IDBStorage(dbName);
+  }
+
+  public get tileStorage(): IDBStorage {
+    return this._tileStorage;
+  }
+
+  public async fetch(url: string, callback: (url: string) => Promise<Response>): Promise<ArrayBuffer> {
```

Comment:
This fetch function now takes a callback function so that it can always return content even if the db request is unsuccessful. For now, I've kept this function as a part of the tile request channel, but it can be moved.

Reply:
This couples things more. Try removing completely IDBTileRequestChannel

Reply:
I decided to keep this fetch function as simple as possible to read since it is the one public function, meaning that instead of having the actual cache searching in this function, I kept the cache searching separate. I think that if the fetch and searchCache functions were combined it would be a much more complicated function to read, but if you think they should be combined, I can certainly do that.